src/hyperfine/parameter_range.rs

```diff
@@ -0,0 +1,176 @@
+use crate::hyperfine::error::ParameterScanError;
+use crate::hyperfine::types::{Command, NumericType};
+use clap::Values;
+use rust_decimal::Decimal;
+use std::ops::{Add, AddAssign, Div, Sub};
+use std::str::FromStr;
+
+trait Numeric:
+    Add<Output = Self>
+    + Sub<Output = Self>
+    + Div<Output = Self>
+    + AddAssign
+    + PartialOrd
+    + Copy
+    + Clone
+    + From<i32>
+    + Into<NumericType>
+{
+}
+impl<
+        T: Add<Output = Self>
+            + Sub<Output = Self>
+            + Div<Output = Self>
+            + AddAssign
+            + PartialOrd
+            + Copy
+            + Clone
+            + From<i32>
+            + Into<NumericType>,
+    > Numeric for T
+{
+}
+
+struct RangeStep<T> {
+    state: T,
+    end: T,
+    step: T,
+}
+
+impl<T: Numeric> RangeStep<T> {
+    fn new(start: T, end: T, step: T) -> RangeStep<T> {
+        RangeStep {
+            state: start,
+            end,
+            step,
+        }
+    }
+}
+
+impl<T: Numeric> Iterator for RangeStep<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.state > self.end {
+            return None;
+        }
+        let return_val = self.state;
+        self.state += self.step;
+
+        Some(return_val)
+    }
+}
+
+fn validate_params<T: Numeric>(start: T, end: T, step: T) -> Result<(), ParameterScanError> {
+    if end < start {
+        return Err(ParameterScanError::EmptyRange);
+    }
+
+    if step == T::from(0) {
+        return Err(ParameterScanError::ZeroStep);
+    }
+
+    const MAX_PARAMETERS: i32 = 100_000;
+    let steps = (end - start + T::from(1)) / step;
+    if steps > T::from(MAX_PARAMETERS) {
+        return Err(ParameterScanError::TooLarge);
+    }
+
+    Ok(())
+}
+
+fn build_parameterized_commands<'a, T: Numeric>(
+    param_min: T,
+    param_max: T,
+    step: T,
+    command_strings: Vec<&'a str>,
+    param_name: &'a str,
+) -> Result<Vec<Command<'a>>, ParameterScanError> {
+    validate_params(param_min, param_max, step)?;
+    let param_range = RangeStep::new(param_min, param_max, step);
+    let mut commands = vec![];
+
+    for value in param_range {
+        for cmd in &command_strings {
+            commands.push(Command::new_parametrized(cmd, param_name, value.into()));
+        }
+    }
+    Ok(commands)
+}
+
+pub fn get_parameterized_commands<'a>(
+    command_strings: Values<'a>,
+    mut vals: clap::Values<'a>,
+    step: Option<&str>,
+) -> Result<Vec<Command<'a>>, ParameterScanError> {
+    let command_strings = command_strings.collect::<Vec<&str>>();
+    let param_name = vals.next().unwrap();
+    let param_min = vals.next().unwrap();
+    let param_max = vals.next().unwrap();
+
+    // attempt to parse as integers
+    if let (Ok(param_min), Ok(param_max)) = (param_min.parse::<i32>(), param_max.parse::<i32>()) {
+        let step = step.unwrap_or("1").parse::<i32>()?;
```

Comment:
This will result in an error if we run
```
--parameter-scan t 0 1 --parameter-step-size 0.1
```

I think we could potentially support this if we add the step-parsing in the `if let` above. This way, all of the parameters would be parsed as decimals below. What do you think?

Reply:
This is a nice catch. I took your suggestion and moved the step parsing line in the `if let` statement.